components/layout/flex.rs

```diff
@@ -69,22 +73,195 @@ impl AxisSize {
+    /// The 'flex-grow' property of this item.
+    pub flex_grow: f32,
+    /// The 'flex-shrink' property of this item.
+    pub flex_shrink: f32,
+    /// The 'order' property of this item.
+    pub order: i32,
+    /// Whether the main size has met its constraint.
+    pub is_freezed: bool,
+    /// True if this flow has property 'visibility::collapse'.
+    pub is_strut: bool
 }
 
 impl FlexItem {
-    fn new(flow: FlowRef) -> FlexItem {
+    pub fn new(flow: FlowRef) -> FlexItem {
+        let style = flow.as_block().fragment.style.clone();
+        let flex_grow = style.get_position().flex_grow;
+        let flex_shrink = style.get_position().flex_shrink;
+        let order = style.get_position().order;
+        // TODO(stshine): for item with visibility:collapse, set is_strut to true.
+
         FlexItem {
-            flow: flow
+            main_size: Au(0),
+            base_size: Au(0),
+            min_size: Au(0),
+            max_size: MAX_AU,
+            flow: flow,
+            style: style,
+            flex_grow: flex_grow,
+            flex_shrink: flex_shrink,
+            order: order,
+            is_freezed: false,
+            is_strut: false
+        }
+    }
+
+    /// Initialize the used flex base size, minimal main size and maximal main size.
+    /// For block mode container this method should be called in assign_block_size()
+    /// pass so that the item has already been layouted.
+    pub fn init_sizes(&mut self, containing_length: Au, mode: Mode) {
+        let block = flow_ref::deref_mut(&mut self.flow).as_mut_block();
+        match mode {
+            // TODO(stshine): the definition of min-{width, height} in style component
+            // should change to LengthOrPercentageOrAuto for automatic implied minimal size.
+            // https://drafts.csswg.org/css-flexbox-1/#min-size-auto
+            Mode::Inline => {
+                let basis = from_flex_basis(self.style.get_position().flex_basis,
+                                            self.style.content_inline_size(),
+                                            Some(containing_length));
+
+                // These methods compute auto margins to zero length, which is exactly what we want.
+                block.fragment.compute_border_and_padding(containing_length,
+                                                          border_collapse::T::separate);
+                block.fragment.compute_inline_direction_margins(containing_length);
+                block.fragment.compute_block_direction_margins(containing_length);
+
+                let adjust_size = match self.style.get_position().box_sizing {
+                    box_sizing::T::border_box => {
+                        let margin = self.style.logical_margin();
+                        (MaybeAuto::from_style(margin.inline_start, Au(0)).specified_or_zero() +
+                         MaybeAuto::from_style(margin.inline_end, Au(0)).specified_or_zero())
+                    }
```

Comment:
Here when the the `box-sizing` property is `border-box`, percentage padding included in main size is calculated to zero as a result of `bubble_inline_sizes`. Maybe I should recalculate them?